File: examples/nodejs/gcFormsApiClient.ts
import axios, { type AxiosInstance } from "axios";
import type {
  EncryptedFormSubmission,
  FormSubmissionProblem,
  NewFormSubmission,
} from "./types.js";

export class GCFormsApiClient {
  private httpClient: AxiosInstance;

  public constructor(apiUrl: string, accessToken: string) {
    this.httpClient = axios.create({
      baseURL: apiUrl,
      timeout: 3000,
      headers: { Authorization: `Bearer ${accessToken}` },
    });
  }

  public getFormTemplate(formId: string): Promise<Record<string, unknown>> {
    return this.httpClient
      .get<Record<string, unknown>>(`/forms/${formId}/template`)
      .then((response) => response.data)
      .catch((error) => {
        throw new Error("Failed to retrieve form template", { cause: error });
      });
  }

  public getNewFormSubmissions(formId: string): Promise<NewFormSubmission[]> {
    return this.httpClient
      .get<NewFormSubmission[]>(`/forms/${formId}/submission/new`)
      .then((response) => response.data)
      .catch((error) => {
        throw new Error("Failed to retrieve new form submissions", {
          cause: error,
        });
      });
  }

  public getFormSubmission(
    formId: string,
    submissionName: string,
  ): Promise<EncryptedFormSubmission> {
    return this.httpClient
      .get<EncryptedFormSubmission>(
        `/forms/${formId}/submission/${submissionName}`,
      )
      .then((response) => response.data)
      .catch((error) => {
        throw new Error("Failed to retrieve form submission", { cause: error });
      });
  }

  public confirmFormSubmission(
    formId: string,
    submissionName: string,
    confirmationCode: string,
  ): Promise<void> {
    return this.httpClient
      .put<void>(
        `/forms/${formId}/submission/${submissionName}/confirm/${confirmationCode}`,
      )
      .then(() => Promise.resolve())
      .catch((error) => {
        throw new Error("Failed to confirm form submission", { cause: error });
      });
  }

  public reportProblemWithFormSubmission(
    formId: string,
    submissionName: string,
    problem: FormSubmissionProblem,
  ): Promise<void> {
    return this.httpClient
      .post<void>(
        `/forms/${formId}/submission/${submissionName}/problem`,
        problem,
        {
          headers: {
            "Content-Type": "application/json",
          },
        },
      )
      .then(() => Promise.resolve())
      .catch((error) => {
        throw new Error("Failed to report problem with form submission", {
          cause: error,
        });
      });
  }
}

File: examples/python/gc_forms_api_client.py
import httpx
from typing import List
from data_structures import (
    NewFormSubmission,
    EncryptedFormSubmission,
    FormSubmissionProblem,
)


class GCFormsApiClient:
    form_id: str
    httpClient: httpx.Client

    def __init__(self, form_id: str, api_url: str, access_token: str):
        self.form_id = form_id
        self.httpClient = httpx.Client(
            base_url=api_url,
            timeout=3,
            headers={"Authorization": f"Bearer {access_token}"},
        )

    def get_form_template(self) -> dict:
        try:
            response = self.httpClient.get(f"/forms/{self.form_id}/template")
            response.raise_for_status()
            return dict(response.json())
        except Exception as exception:
            raise Exception("Failed to retrieve form template") from exception

    def get_new_form_submissions(self) -> List[NewFormSubmission]:
        try:
            response = self.httpClient.get(f"/forms/{self.form_id}/submission/new")
            response.raise_for_status()
            return [NewFormSubmission.from_json(item) for item in response.json()]
        except Exception as exception:
            raise Exception("Failed to retrieve new form submissions") from exception

    def get_form_submission(self, submission_name: str) -> EncryptedFormSubmission:
        try:
            response = self.httpClient.get(
                f"/forms/{self.form_id}/submission/{submission_name}"
            )
            response.raise_for_status()
            return EncryptedFormSubmission.from_json(response.json())
        except Exception as exception:
            raise Exception("Failed to retrieve form submission") from exception

    def confirm_form_submission(
        self, submission_name: str, confirmation_code: str
    ) -> None:
        try:
            response = self.httpClient.put(
                f"/forms/{self.form_id}/submission/{submission_name}/confirm/{confirmation_code}"
            )
            response.raise_for_status()
        except Exception as exception:
            raise Exception("Failed to confirm form submission") from exception

    def report_problem_with_form_submission(
        self, submission_name: str, problem: FormSubmissionProblem
    ) -> None:
        try:
            response = self.httpClient.post(
                f"/forms/{self.form_id}/submission/{submission_name}/problem",
                json=problem.to_json(),
            )
            response.raise_for_status()
        except Exception as exception:
            raise Exception(
                "Failed to report problem with form submission"
            ) from exception
